src/backend/catalog/index.c

```diff
@@ -750,8 +751,16 @@ index_create(Relation heapRelation,
  		 * Index of Table Variables should also be in ENR and the index
 		 * can only be created as part of DECLARE @[name] TABLE;
 		 * when specifying PRIMARY KEYs and/or CONSTRAINTs.
+		 *
+		 * Indexes of temp tables should be in enr as well, to prevent cache issues.
+		 * Index names can actually contain '#' naturally, but because of construct_unique_index_name
+		 * we have no way of checking that without the original index name provided to parser.
 		 */
-		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && indexRelationName[0] == '@');
+		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && (indexRelationName[0] == '@' || strchr(indexRelationName, '#') != NULL));
+
+		/* One last check, if the table has any dependencies */
+		if(CheckTempTableHasDependencies(heapRelation->rd_att))
```

Comment:
I can think of two issues:

1. This isnt completely consistent with the check in heap_create_with_catalog(). This check here doesnt ignore Table Variables but the one in heap_create_with_catalog() ignores Table Variables. The reason is that we want everything associated to TVs in ENR.

2. Consider the following test
-> create table where CheckTempTableHasDependencies returns false this means table is not in ENR
-> Alter table to include a column where CheckTempTableHasDependencies would return True
-> Create an index, now CheckTempTableHasDependencies would return True so you will end up with table not in ENR, and index not in ENR.

Reply:
Perhaps `get_ENR_withoid()` can help